gunrock/app/mf/mf_test.cuh

```diff
@@ -0,0 +1,336 @@
+    std::vector<vertex_descriptor> p(graph.nodes);
+    vertex_descriptor s = vertex(src, g);
+
+    typename property_map<BGraphT, vertex_index_t>::type
+        indexmap = get(vertex_index, g);
+
+    //
+    // Perform Boost reference
+    //
+
+    util::CpuTimer cpu_timer;
+    cpu_timer.Start();
+
+    // TODO: use boost routine to do computation on CPU, e.g.:
+    // dijkstra_shortest_paths(g, s,
+    //        distance_map(boost::make_iterator_property_map(
+    //            d.begin(), get(boost::vertex_index, g))));
+    cpu_timer.Stop();
+    float elapsed = cpu_timer.ElapsedMillis();
+
+    // TODO: extract results on CPU, e.g.:
+    // typedef std::pair<VertexT, ValueT> PairT;
+    // PairT* sort_dist = new PairT[graph.nodes];
+    // typename graph_traits <BGraphT>::vertex_iterator vi, vend;
+    // for (tie(vi, vend) = vertices(g); vi != vend; ++vi)
+    // {
+    //    sort_dist[(*vi)].first  = (*vi);
+    //    sort_dist[(*vi)].second = d[(*vi)];
+    // }
+    // std::stable_sort(
+    //    sort_dist, sort_dist + graph.nodes,
+    //    [](const PairT &a, const PairT &b) -> bool
+    //    {
+    //        return a.first < b.first;
+    //    });
+    // for (VertexT v = 0; v < graph.nodes; ++v)
+    //    distances[v] = sort_dist[v].second;
+    delete[] sort_dist; sort_dist = NULL;
+
+    return elapsed;
+#else
+
+    // TODO: simple cpu reference without boost, e.g.:
+
+    // typedef typename GraphT::VertexT VertexT;
+    // typedef typename GraphT::SizeT   SizeT;
+    // typedef std::pair<VertexT, ValueT> PairT;
+    // struct GreaterT
+    // {
+    //    bool operator()(const PairT& lhs, const PairT& rhs)
+    //    {
+    //        return lhs.second > rhs.second;
+    //    }
+    // };
+    // typedef std::priority_queue<PairT, std::vector<PairT>, GreaterT> PqT;
+    //
+    // for (VertexT v = 0; v < graph.nodes; v++)
+    // {
+    //     distances[v] = util::PreDefinedValues<ValueT>::MaxValue;
+    // }
+    // distances[src] = 0;
+    //
+    // PqT pq;
+    // pq.push(std::make_pair(src, 0));
+    util::CpuTimer cpu_timer;
+    cpu_timer.Start();
+    // while (!pq.empty())
+    // {
+    //    auto pair = pq.top();
+    //    pq.pop();
+    //    VertexT v = pair.first;
+    //    ValueT v_distance = pair.second;
+    //    if (v_distance > distances[v])
+    //        continue;
+    //
+    //    SizeT e_start = graph.GetNeighborListOffset(v);
+    //    SizeT e_end = e_start + graph.GetNeighborListLength(v);
+    //    for (SizeT e = e_start; e < e_end; e++)
+    //    {
+    //        VertexT u = graph.GetEdgeDest(e);
+    //        ValueT u_distance = v_distance + graph.edge_values[e];
+    //        if (u_distance < distances[u])
+    //        {
+    //            distances[u] = u_distance;
+    //            pq.push(std::make_pair(u, u_distance));
+    //        }
+    //    }
+    // }
+    cpu_timer.Stop();
+    float elapsed = cpu_timer.ElapsedMillis();
+
+    return elapsed;
+#endif
```

Comment:
Remove this `#endif` if you end up removing the boost part, and copy paste/implement your cpu version here. Follow the example in sssp if you need a reference.
More or less it is taking your code and making sure it matches gunrock stuff. Don't worry about the cpu timers.